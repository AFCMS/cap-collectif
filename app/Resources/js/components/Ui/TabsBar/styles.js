// @flow
import styled, { css } from 'styled-components';
import colors from '../../../utils/colors';

const verticalMixin = css`
  flex-direction: column;
  align-items: flex-start;
  flex-wrap: nowrap;
`;

export const TabsBarContainer = styled.ul`
  opacity: ${props => (props.show ? 1 : 0)};
  height: 100%;
  width: 100%;
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  flex-direction: row;
  align-items: center;
  margin: 0;
  padding: 0;

  ${props => props.vertical && verticalMixin}
`;

const activeNavItem = css`
  color: ${props => props.theme.mainNavbarTextActive};
  background-color: ${props => props.theme.mainNavbarBgActive};
`;

export const TabsItemContainer = styled.li`
  position: relative;
  display: block;
  width: ${props => (props.vertical ? '100%' : 'auto')};
  height: 100%;
  text-align: center;
  white-space: nowrap;
  list-style: none;
  cursor: pointer;

  ${props => props.active && activeNavItem}

  & > * {
    display: flex;
    flex-direction: row;
    align-items: center;
    text-align: left;
    height: 100%;
    padding: ${props => (props.vertical ? '10px 15px' : '0 15px')};
  }
`;

const activeTabsLinkMixin = css`
  color: ${props => props.theme.mainNavbarTextActive};
  background-color: ${props => props.theme.mainNavbarBgActive};
  text-decoration: none;
`;

const hoverTabsLinkMixin = css`
  color: ${props => props.theme.mainNavbarTextHover};
  background-color: ${props => props.theme.mainNavbarBgActive};
  text-decoration: none;
`;

export const TabsLink = styled.a`
  color: ${props => props.theme.mainNavbarText};
  text-decoration: none;
  cursor: pointer;

  ${props => props.active && activeTabsLinkMixin}

  &:hover,
  &:focus {
    ${hoverTabsLinkMixin}
  }
`;

export const TabsDivider = styled.div`
  height: 1px;
  margin: 9px 0;
  overflow: hidden;
  background-color: #e5e5e5;
`;

export const Dropdown = styled.div`
  position: relative;
  display: block;
  width: 100%;
  height: 100%;
  padding: 0;
`;

const verticalDropdownMenuMixin = css`
  position: relative;
  background: transparent;
  border: 0;
  box-shadow: none;
  width: 100%;
`;

export const DropdownMenu = styled.ul`
  position: absolute;
  top: 100%;
  ${props =>
    props.pullRight
      ? css`
          right: 0;
        `
      : css`
          left: 0;
        `}
  z-index: 1000;
  display: ${props => (props.show ? 'block' : 'none')};
  min-width: 160px;
  padding: 5px 0;
  margin: 0;
  font-size: 14px;
  text-align: left;
  list-style: none;
  background-color: ${props => props.theme.mainNavbarBg};
  border: 1px solid #e3e3e3;
  border-radius: 4px;
  border-top-left-radius: 0;
  border-top-right-radius: 0;
  box-shadow: 0 6px 12px rgba(0, 0, 0, 0.175);

  ${props => props.vertical && verticalDropdownMenuMixin}

  a {
    display: block;
    padding: 3px 20px;
    clear: both;
    font-weight: 400;
    white-space: nowrap;
  }
`;

export const DropdownToggle = styled.button`
  background: transparent;
  border: none;
  text-decoration: none;
  width: 100%;
  height: 100%;
  text-align: left;
  padding: ${props => (props.vertical ? '10px 15px' : '0 15px')};
  color: ${props => props.theme.mainNavbarText};

  &:hover,
  &:focus {
    ${hoverTabsLinkMixin}
  }

  & .caret {
    margin-left: 5px;
  }
`;

export const DropdownSection = styled.ul`
  border-top: 1px solid #e3e3e3;
  border-bottom: 1px solid #e3e3e3;
  margin: 9px 0;
  padding: 9px 0;
`;

export const DropdownSectionTitle = styled.span`
  display: block;
  padding: 3px 20px;
  font-size: 14px;
  color: ${colors.darkGray};
  white-space: nowrap;
`;
