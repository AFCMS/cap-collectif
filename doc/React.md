### How to write localizable components

Just import the appropriate [component](https://github.com/yahoo/react-intl/wiki#the-react-intl-module) from `react-intl`

- For localizable text use
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage).

- For date and time:
[`<FormattedDate>`](https://github.com/yahoo/react-intl/wiki/Components#formatteddate)
[`<FormattedTime>`](https://github.com/yahoo/react-intl/wiki/Components#formattedtime)
[`<FormattedRelative>`](https://github.com/yahoo/react-intl/wiki/Components#formattedrelative)

- For numbers and currencies:
[`<FormattedNumber>`](https://github.com/yahoo/react-intl/wiki/Components#formattednumber)
[`<FormattedPlural>`](https://github.com/yahoo/react-intl/wiki/Components#formattedplural)

- If possible, do not use `<FormattedHTMLMessage>`, see how to use *Rich Text Formatting* with
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage)

- When you need an imperative formatting API, use the [`injectIntl`](https://github.com/yahoo/react-intl/wiki/API#injectintl) High-Order Component.
