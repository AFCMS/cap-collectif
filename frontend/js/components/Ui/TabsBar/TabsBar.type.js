// @flow

export type Item = {|
  +active: boolean,
  +hasEnabledFeature: boolean,
  +id: number,
  +link: string,
  +title: string,
  +children: Item[],
|};
