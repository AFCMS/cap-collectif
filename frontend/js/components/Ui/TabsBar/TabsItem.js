// @flow
import * as React from 'react';
import { type IntlShape } from 'react-intl';

import TabsBarDropdown from './TabsBarDropdown';

import * as S from './styles';

type Props = {|
  intl: IntlShape,
  item: Object,
  vertical: boolean,
|};

class TabsItem extends React.PureComponent<Props> {
  static defaultProps = {
    vertical: false,
  };

  render() {
    const { item, intl, vertical } = this.props;
    const ariaTitle = `${item.title} - ${intl.formatMessage({ id: 'active.page' })}`;

    if (item.hasEnabledFeature) {
      if (item.children && item.children.length > 0) {
        return (
          <TabsBarDropdown
            intl={intl}
            item={item}
            toggleElement={item.title}
            id={`tabsbar-item-${item.id}`}
            vertical={vertical}
          />
        );
      }

      return (
        <S.TabsLink id={`tabs-navbar-link-${item.id}`} href={item.link} active={item.active} title={item.active ? ariaTitle : null}>
          {item.title}
        </S.TabsLink>
      );
    }

    return null;
  }
}

export default TabsItem;
