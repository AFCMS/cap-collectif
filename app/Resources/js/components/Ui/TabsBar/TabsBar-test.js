// @flow
/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import { TabsBar } from './TabsBar';
import { intlMock } from '../../../mocks';

describe('<TabsBar />', () => {

  const props = {
      intl: intlMock,
      items: [],
      overflowEnable: false,
      vertical: false,
  };

  it('should render', () => {
    const wrapper = shallow(
      <TabsBar {...props} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

});
