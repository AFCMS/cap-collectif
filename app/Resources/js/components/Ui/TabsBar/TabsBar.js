// @flow
import * as React from 'react';
import { injectIntl, type IntlShape } from 'react-intl';

import TabsItem from './TabsItem';
import TabsBarDropdown from './TabsBarDropdown';

import * as S from './styles';

type Props = {|
  intl: IntlShape,
  items: Array<Object>,
  /*
   * Whether or not the TabsBar can have an overflow, this should be `true` on mobile devices.
   */
  overflowEnable?: boolean,
  /*
   * Whether the TabsBar is vertical or horizontal, it should be `true` on mobile devices
   */
  vertical?: boolean,
|};

type State = {|
  shouldRender: boolean,
  shouldDisplaySeeMore: boolean,
  overflowIndex: number,
|};

export class TabsBar extends React.Component<Props, State> {
  containerWidth: number;

  seeMoreWidth: number;

  containerRef: Object;

  seeMoreRef: Object;

  itemsWidth: Array<number>;

  static defaultProps = {
    overflowEnable: true,
    vertical: false,
  };

  constructor(props: Props) {
    super(props);

    // Usefull for mobile case
    if (props.overflowEnable) {
      this.state = {
        shouldRender: false,
        shouldDisplaySeeMore: true,
        overflowIndex: props.items.length,
      };
    } else {
      this.state = {
        shouldRender: true,
        shouldDisplaySeeMore: false,
        overflowIndex: props.items.length,
      };
    }

    this.itemsWidth = [];
    this.containerRef = null;
    this.seeMoreRef = null;
  }

  componentDidMount() {
    const { overflowEnable } = this.props;

    if (overflowEnable) {
      this.handleOverflow();
      window.addEventListener('resize', this.handleOverflow);
    }
  }

  componentWillUnmount() {
    const { overflowEnable } = this.props;

    if (overflowEnable) {
      window.removeEventListener('resize', this.handleOverflow);
    }
  }

  setContainerRef = (element: React.Node) => {
    this.containerRef = element;
  };

  setSeeMoreRef = (element: React.Node) => {
    this.seeMoreRef = element;
  };

  getSeeMoreWidth = (): number => {
    if (this.seeMoreRef) {
      const { width } = this.seeMoreRef.getBoundingClientRect();
      return width;
    }

    return 0;
  };

  getContainerWidth = (): number => {
    if (this.containerRef) {
      const { width } = this.containerRef.getBoundingClientRect();
      return width;
    }

    return 0;
  };

  getTotalItemsWidth = (): number => this.itemsWidth.reduce((acc, val) => acc + val, 0);

  handleItemWidth = (element: Object) => {
    if (element) {
      this.itemsWidth.push(element.offsetWidth);
    }
  };

  handleOverflow = () => {
    const { items } = this.props;
    this.seeMoreWidth = !this.seeMoreWidth ? this.getSeeMoreWidth() : this.seeMoreWidth;
    this.containerWidth = this.getContainerWidth();
    const totalItemsWidth = this.getTotalItemsWidth();

    if (totalItemsWidth > this.containerWidth) {
      const overflowIndex = this.getOverflowIndex(this.seeMoreWidth, this.containerWidth);

      this.setState({
        shouldDisplaySeeMore: overflowIndex < items.length,
        overflowIndex,
      });
    } else {
      this.setState({
        shouldDisplaySeeMore: false,
        overflowIndex: items.length,
      });
    }

    this.setState({ shouldRender: true });
  };

  getOverflowIndex = (initWidth: number, maxWidth: number): number => {
    const { items } = this.props;

    if (this.itemsWidth) {
      let counter = initWidth;

      let maxIndex = items.length;

      // Break the loop when accumulation of items width is greater than container size
      // Then, save the index to get the maximum number of items to display
      for (let index = 0; index < maxIndex; index++) {
        counter += this.itemsWidth[index];
        if (counter >= maxWidth) {
          maxIndex = index;
          break;
        }
      }
      return maxIndex;
    }

    return items.length;
  };

  getDisplayedItems = (): Array<Object> => {
    const { items } = this.props;
    const { overflowIndex } = this.state;
    return items.filter((item, index) => index < overflowIndex);
  };

  getOverflowedItems = (): Array<Object> => {
    const { items } = this.props;
    const { overflowIndex } = this.state;
    return items.filter((item, index) => index >= overflowIndex);
  };

  renderSeeMore = () => {
    const { intl, vertical } = this.props;
    const overflowedItems = this.getOverflowedItems();
    const seeMoreItem = {
      title: intl.formatMessage({ id: 'global.navbar.see_more' }),
      children: overflowedItems,
    };

    return (
      // $FlowFixMe ref on a styled component
      <S.TabsItemContainer vertical={vertical} ref={this.setSeeMoreRef}>
        <TabsBarDropdown
          item={seeMoreItem}
          id="tabsbar-dropdown-see-more"
          toggleElement={intl.formatMessage({ id: 'global.navbar.see_more' })}
          intl={intl}
        />
      </S.TabsItemContainer>
    );
  };

  render() {
    const { intl, vertical } = this.props;
    const { shouldDisplaySeeMore, shouldRender } = this.state;
    const displayedItems = this.getDisplayedItems();

    return (
      // $FlowFixMe ref on a styled component
      <S.TabsBarContainer show={shouldRender} vertical={vertical} ref={this.setContainerRef}>
        {displayedItems.map((item, index) => (
          <S.TabsItemContainer key={index} vertical={vertical} ref={this.handleItemWidth}>
            <TabsItem item={item} intl={intl} vertical={vertical} />
          </S.TabsItemContainer>
        ))}
        {shouldDisplaySeeMore && this.renderSeeMore()}
      </S.TabsBarContainer>
    );
  }
}

export default injectIntl(TabsBar);
