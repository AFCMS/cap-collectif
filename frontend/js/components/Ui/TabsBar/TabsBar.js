// @flow
import * as React from 'react';
import { useIntl } from 'react-intl';

import TabsItem from './TabsItem';
import TabsBarDropdown from './TabsBarDropdown';

import * as S from './styles';
import useShowMore from '../../../utils/hooks/useShowMore';

type Props = {|
  items: Array<Object>,
  vertical?: boolean,
|};

const TabsBar = ({ items, vertical }: Props) => {
  const intl = useIntl();
  const [
    containerRef,
    seeMoreRef,
    handleItemWidth,
    overflowIndex,
    shouldDisplaySeeMore,
  ] = useShowMore(!vertical, items.length);

  const renderSeeMore = () => {
    const overflowedItems = items.filter((item, index) => index >= overflowIndex);
    const seeMoreItem = {
      title: intl.formatMessage({ id: 'global.navbar.see_more' }),
      children: overflowedItems,
    };

    return (
      // $FlowFixMe ref on a styled component
      <S.TabsItemContainer vertical={vertical} ref={seeMoreRef}>
        <TabsBarDropdown
          item={seeMoreItem}
          id="tabsbar-dropdown-see-more"
          toggleElement={intl.formatMessage({ id: 'global.navbar.see_more' })}
          intl={intl}
        />
      </S.TabsItemContainer>
    );
  };

  return (
    // $FlowFixMe ref on a styled component
    <S.TabsBarContainer show vertical={vertical} ref={containerRef}>
      {items.map((item, index) => {
        return index < overflowIndex ? (
          <S.TabsItemContainer
            key={index}
            vertical={vertical}
            ref={handleItemWidth}
            className="tabsbar-item-wrapper">
            <TabsItem key={index} item={item} intl={intl} vertical={vertical} />
          </S.TabsItemContainer>
        ) : null;
      })}
      {shouldDisplaySeeMore && renderSeeMore()}
    </S.TabsBarContainer>
  );
};
TabsBar.defaultProps = {
  vertical: false,
};

export default TabsBar;
