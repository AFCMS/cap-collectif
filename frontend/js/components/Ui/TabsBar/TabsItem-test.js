// @flow
/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import TabsItem from './TabsItem';
import { intlMock } from '../../../mocks';

describe('<TabsItem />', () => {
  const props = {
    intl: intlMock,
    item: {},
    vertical: false,
  };

  it('should render', () => {
    const wrapper = shallow(<TabsItem {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
