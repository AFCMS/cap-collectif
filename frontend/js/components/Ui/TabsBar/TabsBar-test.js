// @flow
/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import TabsBar from './TabsBar';

describe('<TabsBar />', () => {
  const props = {
    items: [],
    vertical: false,
  };

  it('should render', () => {
    const wrapper = shallow(<TabsBar {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
