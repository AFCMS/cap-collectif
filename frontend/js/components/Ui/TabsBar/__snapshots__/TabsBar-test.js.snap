// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabsBar /> should render 1`] = `
<styles__TabsBarContainer
  show={true}
  vertical={false}
>
  <styles__TabsItemContainer
    vertical={false}
  >
    <TabsBarDropdown
      id="tabsbar-dropdown-see-more"
      intl={
        Object {
          "formatDate": [Function],
          "formatHTMLMessage": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatPlural": [Function],
          "formatRelative": [Function],
          "formatTime": [Function],
          "formats": Object {},
          "locale": "fr-FR",
          "messages": Object {},
          "now": [Function],
        }
      }
      item={
        Object {
          "children": Array [],
          "title": "global.navbar.see_more",
        }
      }
      pullRight={false}
      toggleElement="global.navbar.see_more"
      vertical={false}
    />
  </styles__TabsItemContainer>
</styles__TabsBarContainer>
`;
